compiler/rustc_hir_typeck/src/closure.rs

```diff
@@ -244,7 +244,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         }
                     }
                 }
-                if inferred_sig.visit_with(&mut MentionsTy { expected_ty }).is_continue() {
+                // FIXME(non_lifetime_binder): Don't infer a signature with late-bound ty/ct vars
+                if inferred_sig.visit_with(&mut MentionsTy { expected_ty }).is_continue()
+                    && !inferred_sig.has_non_region_late_bound()
```

Comment:
is this check not enough to prevent ICEs here? also, can you check that `!inferred_sig.has_placeholders` because inferring a closure signature to contain placeholders is wrong as the closure type is used outside of the binder which created them

Reply:
No, this just makes sure that we just don't eagerly deduce a closure signature from pending obligations.

We can still infer placeholders when we process pending obligations later, so we'd need to do something _after_ typechecking to make sure that the closures don't capture any type placeholders.

Reply:
I agree that check here along with `!inferred_sig.has_placeholders` should be enough. 
We keep getting ICE because of #109505 and probably because we're not using ty vars in the root universe.